pipelines/context.go

```diff
@@ -0,0 +1,66 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package pipelines
+
+import (
+	"context"
+)
+
+type contextKey struct{}
+
+var activePathwayKey = contextKey{}
+
+// ToContext returns a copy of the given context which includes the pathway p.
+func ToContext(ctx context.Context, p Pathway) context.Context {
+	return context.WithValue(ctx, activePathwayKey, p)
+}
+
+// FromContext returns the pathway contained in the given context.
+func FromContext(ctx context.Context) (p Pathway, ok bool) {
+	if ctx == nil {
+		return Pathway{}, false
+	}
+	v := ctx.Value(activePathwayKey)
+	if s, ok := v.(Pathway); ok {
+		return s, true
+	}
+	return Pathway{}, false
+}
+
+// SetCheckpoint sets a checkpoint on the pathway in the context.
+// if there is no pathway in the context, it creates a new one.
+func SetCheckpoint(ctx context.Context, edge string) (Pathway, context.Context) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	p, ok := FromContext(ctx)
+	if ok {
+		p = p.SetCheckpoint(edge)
+	} else {
+		// skip the edge if there is nothing before this node.
+		p = NewPathway()
+	}
+	ctx = ToContext(ctx, p)
+	return p, ctx
+}
+
+// MergeContexts returns the first context which includes the pathway resulting from merging the pathways
+// contained in all contexts.
+func MergeContexts(ctxs ...context.Context) context.Context {
```

Comment:
What is this used for?

This is an odd use case. I can't imagine someone wanting to ditch multiple contexts and keep only one. The other contexts could easily have valuable information that needs to be propagated further in the application. Then again, I might also be missing context on this myself (pun intended 🙂)

I'd also argue that this function does not need to be part of the package or API because a user could easily write it themselves, in their own app, using the available API.

Reply:
I put a comment here. So The idea is to use this function during fan-in situations (multiple payloads converge into one).
In that case, you can only have 1 resulting context.
Ideally, we would want to merge multiple pathways into 1, that is a combination of all the pathways.
However, right now, we don't have a good way to do that, so we randomly select a Pathway (sampling)

Reply:
I understand. Thanks for explaining. I still don't see the need for this in the API. This can be implemented by users without having this function, since all of the internal calls are to exported functions.